File: docs/absorbed-from-PhenoMCP/bindings/csharp/src/RealClient.cs
using System.Net.Http.Json;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Phenotype.MCP;

/// <summary>
/// Real C# MCP client with HttpClient
/// </summary>
public class RealMCPClient : IDisposable
{
    private readonly HttpClient _http;
    private readonly string _endpoint;
    private string? _sessionId;

    public RealMCPClient(string endpoint)
    {
        _endpoint = endpoint;
        _http = new HttpClient
        {
            Timeout = TimeSpan.FromSeconds(30)
        };
    }

    /// <summary>Initialize with MCP server</summary>
    public async Task<ServerInfo> InitializeAsync()
    {
        var request = new JsonRpcRequest
        {
            JsonRpc = "2.0",
            Id = Guid.NewGuid().ToString(),
            Method = "initialize",
            Params = new Dictionary<string, object>
            {
                ["protocolVersion"] = "2024-11-05",
                ["capabilities"] = new Dictionary<string, object>(),
                ["clientInfo"] = new Dictionary<string, string>
                {
                    ["name"] = "pheno-csharp",
                    ["version"] = "0.1.0"
                }
            }
        };

        var response = await SendAsync<InitializeResponse>(request);
        _sessionId = Guid.NewGuid().ToString();
        return response.Result!.ServerInfo;
    }

    /// <summary>List available tools</summary>
    public async Task<List<McpTool>> ListToolsAsync()
    {
        var request = new JsonRpcRequest
        {
            JsonRpc = "2.0",
            Id = Guid.NewGuid().ToString(),
            Method = "tools/list"
        };

        var response = await SendAsync<ToolsListResponse>(request);
        return response.Result?.Tools ?? new List<McpTool>();
    }

    /// <summary>Call a tool</summary>
    public async Task<ToolResult> CallToolAsync(string name, Dictionary<string, object> args)
    {
        var request = new JsonRpcRequest
        {
            JsonRpc = "2.0",
            Id = Guid.NewGuid().ToString(),
            Method = "tools/call",
            Params = new Dictionary<string, object>
            {
                ["name"] = name,
                ["arguments"] = args
            }
        };

        var response = await SendAsync<ToolCallResponse>(request);
        return response.Result!.Content.First();
    }

    private async Task<T> SendAsync<T>(JsonRpcRequest request) where T : class
    {
        var json = JsonSerializer.Serialize(request);
        var content = new StringContent(json, Encoding.UTF8, "application/json");
        
        var httpResponse = await _http.PostAsync(_endpoint, content);
        var responseJson = await httpResponse.Content.ReadAsStringAsync();
        
        return JsonSerializer.Deserialize<T>(responseJson)!;
    }

    public void Dispose()
    {
        _http.Dispose();
    }
}

// Request/Response types
public class JsonRpcRequest
{
    [JsonPropertyName("jsonrpc")] public string JsonRpc { get; set; } = "2.0";
    [JsonPropertyName("id")] public string Id { get; set; } = "";
    [JsonPropertyName("method")] public string Method { get; set; } = "";
    [JsonPropertyName("params")] public object? Params { get; set; }
}

public class JsonRpcResponse<T> where T : class
{
    [JsonPropertyName("jsonrpc")] public string JsonRpc { get; set; } = "";
    [JsonPropertyName("id")] public string? Id { get; set; }
    [JsonPropertyName("result")] public T? Result { get; set; }
    [JsonPropertyName("error")] public JsonRpcError? Error { get; set; }
}

public class JsonRpcError
{
    [JsonPropertyName("code")] public int Code { get; set; }
    [JsonPropertyName("message")] public string Message { get; set; } = "";
}

public class InitializeResponse
{
    [JsonPropertyName("result")] public InitializeResult? Result { get; set; }
}

public class InitializeResult
{
    [JsonPropertyName("serverInfo")] public ServerInfo ServerInfo { get; set; } = new();
}

public class ServerInfo
{
    [JsonPropertyName("name")] public string Name { get; set; } = "";
    [JsonPropertyName("version")] public string Version { get; set; } = "";
}

public class ToolsListResponse
{
    [JsonPropertyName("result")] public ToolsListResult? Result { get; set; }
}

public class ToolsListResult
{
    [JsonPropertyName("tools")] public List<McpTool> Tools { get; set; } = new();
}

public class McpTool
{
    [JsonPropertyName("name")] public string Name { get; set; } = "";
    [JsonPropertyName("description")] public string Description { get; set; } = "";
}

public class ToolCallResponse
{
    [JsonPropertyName("result")] public ToolCallResult? Result { get; set; }
}

public class ToolCallResult
{
    [JsonPropertyName("content")] public List<ToolResult> Content { get; set; } = new();
}

public class ToolResult
{
    [JsonPropertyName("type")] public string Type { get; set; } = "";
    [JsonPropertyName("text")] public string? Text { get; set; }
}

File: docs/absorbed-from-PhenoMCP/bindings/swift/Sources/PhenoMCP/RealClient.swift
import Foundation

/// Real Swift MCP Client with actual HTTP networking
public actor RealMCPClient {
    private let endpoint: URL
    private let session: URLSession
    private var sessionId: String?
    private var protocolVersion: String = "2024-11-05"

    public init(endpoint: String) throws {
        guard let url = URL(string: endpoint) else {
            throw MCPClientError.invalidEndpoint
        }
        self.endpoint = url
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        self.session = URLSession(configuration: config)
    }

    /// Initialize connection
    public func initialize() async throws -> ServerInfo {
        let body: [String: Any] = [
            "jsonrpc": "2.0",
            "id": UUID().uuidString,
            "method": "initialize",
            "params": [
                "protocolVersion": protocolVersion,
                "capabilities": [:],
                "clientInfo": ["name": "pheno-swift", "version": "0.1.0"]
            ]
        ]
        
        let data = try await post(body)
        guard let result = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let resultObj = result["result"] as? [String: Any],
              let serverInfo = resultObj["serverInfo"] as? [String: Any] else {
            throw MCPClientError.noResult
        }
        
        return ServerInfo(
            name: serverInfo["name"] as? String ?? "",
            version: serverInfo["version"] as? String ?? ""
        )
    }

    /// List tools
    public func listTools() async throws -> [McpTool] {
        let body: [String: Any] = [
            "jsonrpc": "2.0",
            "id": UUID().uuidString,
            "method": "tools/list"
        ]
        
        let data = try await post(body)
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let result = json["result"] as? [String: Any],
              let toolsArray = result["tools"] as? [[String: Any]] else {
            return []
        }
        
        return toolsArray.map { tool in
            McpTool(
                name: tool["name"] as? String ?? "",
                description: tool["description"] as? String ?? "",
                inputSchema: [:]
            )
        }
    }

    /// Call tool
    public func callTool(name: String, arguments: [String: Any]) async throws -> ToolResult {
        let body: [String: Any] = [
            "jsonrpc": "2.0",
            "id": UUID().uuidString,
            "method": "tools/call",
            "params": ["name": name, "arguments": arguments]
        ]
        
        let data = try await post(body)
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let result = json["result"] as? [String: Any],
              let contentArray = result["content"] as? [[String: Any]] else {
            throw MCPClientError.noResult
        }
        
        let first = contentArray.first
        return ToolResult(
            type: first?["type"] as? String ?? "text",
            text: first?["text"] as? String
        )
    }

    private func post(_ body: [String: Any]) async throws -> Data {
        var request = URLRequest(url: endpoint)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        
        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            throw MCPClientError.httpError((response as? HTTPURLResponse)?.statusCode ?? 0)
        }
        
        return data
    }
}

public struct ServerInfo {
    public let name: String
    public let version: String
}

public struct McpTool: Identifiable {
    public let id: String
    public let name: String
    public let description: String
    public let inputSchema: [String: Any]
    
    public init(name: String, description: String, inputSchema: [String: Any]) {
        self.id = name
        self.name = name
        self.description = description
        self.inputSchema = inputSchema
    }
}

public struct ToolResult {
    public let type: String
    public let text: String?
}

public enum MCPClientError: Error {
    case invalidEndpoint
    case noResult
    case invalidResponse
    case httpError(Int)
}
